s3/s3.go

```diff
@@ -151,78 +167,98 @@ func (b *Bucket) Get(path string) (data []byte, err error) {
 	if err != nil {
 		return nil, err
 	}
-	data, err = ioutil.ReadAll(body)
-	body.Close()
-	return data, err
+	defer body.Close()
+
+	return ioutil.ReadAll(body)
 }
 
-// GetReader retrieves an object from an S3 bucket.
-// It is the caller's responsibility to call Close on rc when
-// finished reading.
+// GetReader retrieves an object from an S3 bucket. It is the caller's
+// responsibility to call Close on rc when finished reading.
 func (b *Bucket) GetReader(path string) (rc io.ReadCloser, err error) {
-	req := &request{
-		bucket: b.Name,
-		path:   path,
+
+	req, err := http.NewRequest("GET", b.Region.ResolveS3BucketEndpoint(b.Name), nil)
+	if err != nil {
+		return nil, err
 	}
-	err = b.S3.prepare(req)
+	req.Close = true
+	req.URL.Path += path
+
+	addAmazonDateHeader(req.Header)
+
+	if err := b.S3.Sign(req, b.Auth); err != nil {
+		return nil, err
+	}
+
+	resp, err := requestRetryLoop(req, attempts)
 	if err != nil {
 		return nil, err
 	}
-	for attempt := attempts.Start(); attempt.Next(); {
-		hresp, err := b.S3.run(req)
-		if shouldRetry(err) && attempt.HasNext() {
-			continue
-		}
-		if err != nil {
-			return nil, err
-		}
-		return hresp.Body, nil
+	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusNoContent {
+		return nil, buildError(resp)
 	}
-	panic("unreachable")
+	return resp.Body, nil
 }
 
 // Put inserts an object into the S3 bucket.
```

Comment:
Or actually, just drop the Put doc comment and body only. rather than PutReader's doc comment as well.